### ISSUE TITLE
Questions for a PR to pie! (in AbstractPlotting)

### ISSUE THREAD

**Issue by bvilmann** [JuliaPlots/Makie.jl#782]
Hi!

I am trying to prepare a PR for the `pie!()` recipe in AbstractPlotting. I have found several things that can be optimized. However, I have some questions before finalizing the PR.

To the point.

This code:
```
pie!(scene,1:4, color=1:4, radius = 5, inner_radius=7, offset=3)
```
Plots this:

![image](https://user-images.githubusercontent.com/50515426/101930306-42863b00-3bd8-11eb-9772-8afb3d03f430.png)

So, after looking at it for a bit, I realized: 
It is ellipsoid, the inner_radius is larger than the radius, and it is not offset in the way I expected it to be.

**Content of the pull request**
- `offset`
  - I want to change the name to `rotate` since that is what it does.
  - I think the default input value for `rotate` should be in degrees
  - The `offset` variable should still be optional but be a tuple of `(x,y)` that translates the pie chart in the scene.
- `inner_radius`
  - I believe this should be a value relative to the radius and thereby be limited between 0 and 1. It is nonsensical that the "inner" radius is bigger than the actual radius.
  - This can be added by a "validation" control flow
- `scene` scaling
  - It would be nice if it were possible to scale the pie chart relative to the axes scaling ratio so that it always will be round and not ellipsoid if the axis scaling ratio `!=1`.

**Questions**
- Is there anything above you disagree with or have inquiries for?
- I am quite new to Makie (and Julia). Is there a nice way to call the axis dimensions of the current scene so that it is possible to get a "scaling ratio"?
- ... I am also new to github and I managed to link Github to a local directory (with Github desktop) - but I don't know how I freaking ****ing load the local directory into REPL and test it. How can I test the code and is there a best practice for this? If so - how can I do it? I am using `Atom` and `Github Desktop`.

Thanks for reading (and hopefully responding),
All the best,
B

**Comment by ffreyer** [comment]
> I am quite new to Makie (and Julia). Is there a nice way to call the axis dimensions of the current scene so that it is possible to get a "scaling ratio"?

I think there is a `resolution` for each scene, which gives you a pixel-sized Rect2D. I'm not aware of an inner size for 2D axis, but you probably don't want a pie chart with axis in the background anyway...? And 2D axis dependent things may also break/get deprecated as MakieLayout gets merged more directly.

> ... I am also new to Github and I managed to link Github to a local directory (with Github desktop) - but I don't know how I freaking ****ing load the local directory into REPL and test it. How can I test the code and is there a best practice for this? If so - how can I do it? I am using Atom and Github Desktop.

I think usually people `dev` a package, work on their things on a branch and switch back to master if they need to. With that you can just do `using AbstractPlotting` anywhere and test your changes. You can switch branches from atom, if you open the respective folder (`~/.julia/dev/AbstractPlotting` on linux). If you don't want to do this you need to `include(".../AbstractPlotting.jl"); using .AbstractPlotting` iirc.

**Comment by jkrumbiegel** [comment]
> ... I am also new to Github and I managed to link Github to a local directory (with Github desktop) - but I don't know how I freaking ****ing load the local directory into REPL and test it. How can I test the code and is there a best practice for this? If so - how can I do it? I am using Atom and Github Desktop.

I usually do this:

- open a new vscode / atom window with the folder of the package I want to edit as the working directory
- start the REPL, in some editors it will now already have the environment of the package activated
- if not, do `] activate .` and the current working directory (the package folder) will be activated
- now you can do `using ThePackage` and execute stuff in the repl, edit files (with Revise loaded beforehand) and reexecute stuff with the edits

The other option is to locally install the package into another environment, for example if you want to test its use with other packages you have installed somewhere. Then the steps are

- open folder with the environment you want to use
- again `] activate .` if not already activated
- now install the package of choice into this environment, two options
  - package exists somewhere locally, then do `]dev path/to/local/package`
  - you want to put it in the default `.julia/dev` folder, just do `]dev PackageName`
  - you want to put a local copy into your current folder, do `]dev --local PackageName`

**Comment by jkrumbiegel** [comment]
regarding the scaling, use a MakieLayout LAxis and set `aspect = DataAspect()` or set `autolimitaspect = 1` to have both axes scaled the same, then the pie will be round

**Comment by bvilmann** [comment]
Thanks for the hints! I did not have luck with changing branches. It worked to open the dir and work from there - I have to test the first approach some more! 👍 

@ffreyer & @jkrumbiegel , regarding the autolimitaspect and axis background. I agree: You want the `autolimitaspect = 1` and no axis when calling `pie()`. However, I think it could be nice if it is possible to keep the aspect ratio and the axis when adding a pie chart on top of whatever plot by a convenience function, `pie!()`. This could leave a higher degree of freedom for the user :)

Troubles:
- How do I make the axis disappear? `show_axis = false` did not work
- Either `aspect = DataAspect()` or `autolimitaspect = 1` seems to be working.

Qs:
- How do I catch the axis? By `plot.x` ?
- Do we have any function that can check if the function is invoked by a convenience function?
- Is there a standard way to "hide decoration" and grid without using a lot of lines?
- Do you know if there is a neat way to reload a package so I should not exit julia and precompile everything again and again 🙈 

The suggested changes seems to be working and are in place, however:
![image](https://user-images.githubusercontent.com/50515426/101954992-51331900-3bfd-11eb-83b1-ea4314c03a04.png)

**Comment by bvilmann** [comment]
So, I need a little help:

[Documentation on LAxis ](https://makie.juliaplots.org/dev/makielayout/laxis.html) says that you invoke `LAxis` like this:
```
scene, layout = layoutscene(resolution = (1200, 900))

ax = layout[1, 1] = LAxis(scene, xlabel = "x label", ylabel = "y label",
    title = "Title")
```

We have the scene from the recipe macro:
```
@recipe(Pie, values) do scene
    Theme(
        normalize = true,
        color = :gray,
        strokecolor = :black,
        strokewidth = 1,
        vertex_per_deg = 1,
        radius = 1,
        inner_radius = 0,           # relative inner radius: Float ∈ [0; 1]
        rotate = 0,                 # rotate in degrees
        offset = (0,0),             # Offset x, y
    )
end
```
And in the beginning of the input to the `pie()` function we have this with my additions:
```
function plot!(plot::Pie)

    values = plot[1]

    ax = LAxis(plot)                 # This is returning an error with both 'plot' and 'scene' as arguments
    autolimitaspect = 1             
    aspect = DataAspect()           
    show_axis = false
```

**Summarized:** 
- I don't know where to put your suggestions. They don't make any errors as they are placed there (which is kind of positive?) and I can't find a function where they seem to fit into.

- How do I call the LAxis? I must be honest. After 1 month of Julia coding, I have not quite fully understood the Do-Block syntax when it is matched with a macro and I still have to learn the fuzz about the `Observables`.

I hope you can see where I am going and got lost :)

All the best,
B


****

PS: The `Theme` method in the recipe is called `Attributes` in most of the other plots. Is this a mistake?

I found in `scenes.jl`:

```
    "The plots contained in the Scene."
    plots::Vector{AbstractPlot}
    # TODO why 2?
    theme::Attributes

    attributes::Attributes

```
and in the `Scene()` function:

```
        theme::Attributes, # the default values a scene owns
        attributes::Attributes, # the actual attribute values of a scene
```
I guess `Theme` should be `Attributes` instead since it is designed no to be default values?

**Comment by ffreyer** [comment]
I don't think you can controls axis from inside a recipe/`plot` function. It's something a user can do, from the outside.

**Comment by bvilmann** [comment]
Thanks for your answer! What attribute belongs to the axis size? I tried to look via `fieldnames(typeof(scene))` without any luck :)

**Comment by bvilmann** [comment]
@ffreyer, @jkrumbiegel,

I want to hand in the PR with the changes I was able to implement. I cannot fix the scaling as is because I don't know how to either grab the `scene` or create a `LAxis()` in the plot function for `pie()`.

However, I do have a problem with the test.

After implementing a control flow, which I think is fundamental, I get this:
```
[ Info: Test: Hollow pie chart didn't pass
Record tests: Test Failed at C:\Users\Benjamin\Documents\GitHub\AbstractPlotting.jl\test\ReferenceTests\src\runtests.jl:103
  Expression: false
Stacktrace:
 [1] macro expansion at C:\Users\Benjamin\Documents\GitHub\AbstractPlotting.jl\test\ReferenceTests\src\runtests.jl:103 [inlined]
 [2] macro expansion at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\Test\src\Test.jl:1115 [inlined]
 [3] record_tests(::Dict{String,ReferenceTests.Entry}; recording_dir::String) at C:\Users\Benjamin\Documents\GitHub\AbstractPlotting.jl\test\ReferenceTests\src\runtests.jl:80
inner_radius is a scalar to radius and should be between 0 and 1
```
It fails the test for a hollow pie chart. It seems like I cannot make throw an error inside the plot function. What is the general approach to design these control flows for validating user inputs within this package?

NB: when I try with the master `GLMakie` package I get:
```
julia> pie()
ERROR: Not implemented for pie. You might want to put:  `using Makie` into your code!
```
So it seems like it in both cases is not possible to call a "hollow" pie chart?

All the best,
Benjamin

**Comment by jkrumbiegel** [comment]
Hi Benjamin,

Right now the recipes don't work how I think you understand them to work (closer in spirit to Plots.jl). They are really relatively low-level visual primitives, that sometimes are a bit more complex when you stack a few of them together. But right now, the plot recipes are not a good place to do anything about axis parameters etc. That needs to be handled in a different place, and we're working on different ideas in that regard. For example, each plot recipe that also creates an axis implicitly could have axis parameters set in an `axis` keyword, but this kind of pipeline has to be written globally for all recipes together. Basically, just leave the axis stuff out of the recipe and have the user set the axis according to their whishes, otherwise the interactions get quite complex (what if multiple plots are used together in one axis and each tries to set conflicting parameters to the axis)

**Comment by bvilmann** [comment]
Yes :) It became also my impression that there need to be implemented some new infrastructure to make this happen.

Do you have any comments on the error or should I just make the PR as is?
